README.md
# Detect Recovery

## 描述

这是一款基于move_base的恢复行为插件，可以有效应用于ROS机器人的导航恢复；

## 声明

个人知识产权，禁止商用转载。

`2020/10/27`

**copyright@anchuanxu**


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package detect_recovery
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.1.0 (2020-09-02)
-------------------
* Package Init
-------------------
0.2.0 (2020-10-27)
-------------------
This update content:

* Change the way laser obtains data, making it more flexible to adapt to any laser radar;

* add and cancel function modules;

* parameterized setting frame;

* The processing logic adopts a new way, and every recovery will traverse around 360,degrees to find a direction that can go straight for 20cm.


src/smart_recovery.cpp
/*********************************************************************
* Software License Agreement (BSD License)
* Author: anchuanxu
*********************************************************************/
#include <smart_recovery/smart_recovery.h>
#include <pluginlib/class_list_macros.h>
#include <nav_core/parameter_magic.h>
#include <tf2/utils.h>
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Point.h>
#include <angles/angles.h>
#include <algorithm>
#include <string>

// register this planner as a RecoveryBehavior plugin
PLUGINLIB_EXPORT_CLASS(smart_recovery::SmartRecovery, nav_core::RecoveryBehavior)

namespace smart_recovery
{
SmartRecovery::SmartRecovery(): local_costmap_(NULL), initialized_(false), world_model_(NULL)
{
}

void SmartRecovery::initialize(std::string name, tf2_ros::Buffer*,
                                costmap_2d::Costmap2DROS*, costmap_2d::Costmap2DROS* local_costmap)
{
  if (!initialized_)
  {
    local_costmap_ = local_costmap;

    // get some parameters from the parameter server
    ros::NodeHandle private_nh("~/" + name);
    ros::NodeHandle blp_nh("~/TrajectoryPlannerROS");

    //获取激光数据
    ros::NodeHandle scan_nh;
    ros::Subscriber scan_sub = scan_nh.subscribe<sensor_msgs::LaserScan>("/scan", 1, boost::bind(&SmartRecovery::scanCallback),this,1);

    // we'll simulate every degree by default
    private_nh.param("sim_granularity", sim_granularity_, 0.017);
    private_nh.param("frequency", frequency_, 20.0);

    acc_lim_th_ = nav_core::loadParameterWithDeprecation(blp_nh, "acc_lim_theta", "acc_lim_th", 3.2);
    max_rotational_vel_ = nav_core::loadParameterWithDeprecation(blp_nh, "max_vel_theta", "max_rotational_vel", 1.0);
    min_rotational_vel_ = nav_core::loadParameterWithDeprecation(blp_nh, "min_in_place_vel_theta", "min_in_place_rotational_vel", 0.4);
    blp_nh.param("yaw_goal_tolerance", tolerance_, 0.10);

    world_model_ = new base_local_planner::CostmapModel(*local_costmap_->getCostmap());

    initialized_ = true;
    canGofoword = false;
    canGoback = false;
    pi = 3.14;
  }
  else
  {
    ROS_ERROR("You should not call initialize twice on this object, doing nothing");
  }
}

SmartRecovery::~SmartRecovery()
{
  delete world_model_;
}

void SmartRecovery::runBehavior()
{
  ROS_WARN("------Into SmartRecovery------");

  //判断有没有全局路径规划出来
  std::vector<geometry_msgs::PoseStamped> global_plan;
  if(!global_plan.empty())
  {
      //存在全局规划路径，尝试进行前后挪移
      int turnCount = 0;
      while(turnCount < 4)
      {

          //前方可以行进
          if(canGofoword)
          {
              goStraight(0.05); //前进5cm
              publishZeroVelocity();
          }
          //前方不可前进
          else {
              if(canGoback)
              {
                goBack(0.05);//后退5cm
                publishZeroVelocity();
              }
              else { //不能前进也不能后退，旋转90度试试别的方位
                  turnAngle(pi / 2);
                  turnCount ++;
              }
          }
      }
  }
  else
  {
      //没有全局规划路径，旋转180度，尝试能否给出规划
      rotate180();
  }
}
void SmartRecovery::goStraight(double distance)
{
    double linear_speed = 0.2;
    double linear_duration = distance / linear_speed;
    ros::Rate r(10);
    int ticks = int(linear_duration * 10);
    geometry_msgs::Twist cmd_vel;
    for(int i = 0; i < ticks; i++)
    {
       cmd_vel.linear.x = linear_speed;
       vel_pub_->publish(cmd_vel);
       r.sleep();
    }
    return ;
}
void SmartRecovery::goBack(double distance)
{
    double linear_speed = -0.2;
    double linear_duration = distance / fabs(linear_speed);
    ros::Rate r(10);
    int ticks = int(linear_duration * 10);
    geometry_msgs::Twist cmd_vel;
    for(int i = 0; i < ticks; i++)
    {
       cmd_vel.linear.x = linear_speed;
       vel_pub_->publish(cmd_vel);
       r.sleep();
    }
    return ;
}
void SmartRecovery::publishZeroVelocity()
{
    geometry_msgs::Twist cmd_vel;
    cmd_vel.linear.x = 0.0;
    cmd_vel.linear.y = 0.0;
    cmd_vel.angular.z = 0.0;
    vel_pub_->publish(cmd_vel);
}
void SmartRecovery::scanCallback(const sensor_msgs::LaserScan::ConstPtr msg)
{
    int obs_count = 0;
    int obs_back_count = 0;
    for (int i = 0; i < 180; i++) {
      if (0.1 < msg->ranges[i] && msg->ranges[i] < 0.5) {
        obs_count++;
      }
    }
    for (int i = 540; i < 720; i++) {
      if (0.1 < msg->ranges[i] && msg->ranges[i] < 0.5) {
        obs_count++;
      }
    }
    for (int i = 180; i < 540; i++) {
        if (0.1 < msg->ranges[i] && msg->ranges[i] < 0.5) {
          obs_back_count++;
        }
    }
    if(obs_count > 20)
        canGofoword = false;
    else {
        canGofoword = true;
    }
    if(obs_back_count > 20)
        canGoback = false;
    else {
        canGoback = true;
    }
}
bool SmartRecovery::turnAngle(double angle)
{
    double angle_speed = 0.2;
    double angle_duration = angle / angle_speed;
    ros::Rate r(10);
    int ticks = int(angle_duration * 10);
    geometry_msgs::Twist cmd_vel;
    for(int i = 0; i < ticks; i++)
    {
       cmd_vel.linear.z = angle_speed;
       vel_pub_->publish(cmd_vel);
       r.sleep();
    }
    return true;
}
bool SmartRecovery::turnToGlobalPath(){
//TODO
}
bool SmartRecovery::rotate180()
{
    if (!initialized_)
    {
      ROS_ERROR("This object must be initialized before runBehavior is called");
      return false;
    }

    if (local_costmap_ == NULL)
    {
      ROS_ERROR("The costmap passed to the SmartRecovery object cannot be NULL. Doing nothing.");
      return false;
    }
    ROS_WARN("Smart recovery behavior started.");

    ros::Rate r(frequency_);
    ros::NodeHandle n;
    ros::Publisher vel_pub = n.advertise<geometry_msgs::Twist>("cmd_vel", 10);

    geometry_msgs::PoseStamped global_pose;
    local_costmap_->getRobotPose(global_pose);

    double current_angle = tf2::getYaw(global_pose.pose.orientation);
    double start_angle = current_angle;

    bool got_180 = false;

    while (n.ok() &&
           (!got_180 ||
            std::fabs(angles::shortest_angular_distance(current_angle, start_angle)) > tolerance_))
    {
      // Update Current Angle
      local_costmap_->getRobotPose(global_pose);
      current_angle = tf2::getYaw(global_pose.pose.orientation);

      // compute the distance left to Smart
      double dist_left;
      if (!got_180)
      {
        // If we haven't hit 180 yet, we need to Smart a half circle plus the distance to the 180 point
        double distance_to_180 = std::fabs(angles::shortest_angular_distance(current_angle, start_angle + M_PI));
        dist_left = M_PI + distance_to_180;

        if (distance_to_180 < tolerance_)
        {
          got_180 = true;
        }
      }
      else
      {
        // If we have hit the 180, we just have the distance back to the start
        dist_left = std::fabs(angles::shortest_angular_distance(current_angle, start_angle));
      }

      double x = global_pose.pose.position.x, y = global_pose.pose.position.y;

      // check if that velocity is legal by forward simulating
      double sim_angle = 0.0;
      while (sim_angle < dist_left)
      {
        double theta = current_angle + sim_angle;

        // make sure that the point is legal, if it isn't... we'll abort
        double footprint_cost = world_model_->footprintCost(x, y, theta, local_costmap_->getRobotFootprint(), 0.0, 0.0);
        if (footprint_cost < 0.0)
        {
          ROS_ERROR("Smart recovery can't run in place because there is a potential collision. Cost: %.2f",
                    footprint_cost);
          return false;
        }

        sim_angle += sim_granularity_;
      }

      // compute the velocity that will let us stop by the time we reach the goal
      double vel = sqrt(2 * acc_lim_th_ * dist_left);

      // make sure that this velocity falls within the specified limits
      vel = std::min(std::max(vel, min_rotational_vel_), max_rotational_vel_);

      geometry_msgs::Twist cmd_vel;
      cmd_vel.linear.x = 0.0;
      cmd_vel.linear.y = 0.0;
      cmd_vel.angular.z = vel;

      vel_pub.publish(cmd_vel);

      r.sleep();
    }
    return true;
}
};  // namespace smart_recovery


include/smart_recovery/smart_recovery.h
/*********************************************************************
* Software License Agreement (BSD License)
* Author: anchuanxu
*********************************************************************/
#ifndef SMART_RECOVERY_SMART_RECOVERY_H
#define SMART_RECOVERY_SMART_RECOVERY_H
#include <nav_core/recovery_behavior.h>
#include <costmap_2d/costmap_2d_ros.h>
#include <tf2_ros/buffer.h>
#include <base_local_planner/costmap_model.h>
#include <string>
#include <sensor_msgs/LaserScan.h>

namespace smart_recovery
{
/**
 * @class SmartRecovery
 * @brief A recovery behavior that to relieve the plight of the robot
 */
class SmartRecovery : public nav_core::RecoveryBehavior
{
public:
  /**
   * @brief  Constructor, make sure to call initialize in addition to actually initialize the object
   */
  SmartRecovery();

  /**
   * @brief  Initialization function for the SmartRecovery recovery behavior
   * @param name Namespace used in initialization
   * @param tf (unused)
   * @param global_costmap (unused)
   * @param local_costmap A pointer to the local_costmap used by the navigation stack
   */
  void initialize(std::string name, tf2_ros::Buffer*,
                  costmap_2d::Costmap2DROS*, costmap_2d::Costmap2DROS* local_costmap);

  /**
   * @brief  Run the SmartRecovery recovery behavior.
   */
  void runBehavior();

  /**
   * @brief  Destructor for the smart recovery behavior
   */
  ~SmartRecovery();

private:
  costmap_2d::Costmap2DROS* local_costmap_;
  bool initialized_;
  double sim_granularity_, min_rotational_vel_, max_rotational_vel_, acc_lim_th_, tolerance_, frequency_;
  base_local_planner::CostmapModel* world_model_;
  ros::Publisher* vel_pub_;
  bool canGofoword,canGoback;
  double pi;
  bool rotate180();
  bool turnToGlobalPath();
  bool turnAngle(double angle);
  void scanCallback(const sensor_msgs::LaserScan::ConstPtr msg);
  void publishZeroVelocity();
  void goStraight(double distance);
  void goBack(double distance);
};
};  // namespace smart_recovery
#endif  // SMART_RECOVERY_SMART_RECOVERY_H


include/detect_recovery/detect_recovery.h
/*********************************************************************
*
* Software License Agreement (BSD License)
*
*  Copyright (c) 2020, Anchuanxu.
*  All rights reserved.
* Author: Chuanxu An
*********************************************************************/
#ifndef DETECT_RECOVERY_DETECT_RECOVERY_H
#define DETECT_RECOVERY_DETECT_RECOVERY_H
#include <nav_core/recovery_behavior.h>
#include <costmap_2d/costmap_2d_ros.h>
#include <tf2_ros/buffer.h>
#include <base_local_planner/costmap_model.h>
#include <string>
#include <sensor_msgs/LaserScan.h>

namespace detect_recovery
{
/**
 * @class RoRecovery
 * @brief A recovery behavior that ros the robot in-place to attempt to clear out space
 */
class DetectRecovery : public nav_core::RecoveryBehavior
{
public:
  /**
   * @brief  Constructor, make sure to call initialize in addition to actually initialize the object
   */
  DetectRecovery();

  /**
   * @brief  Initialization function for the DetectRecovery recovery behavior
   * @param name Namespace used in initialization
   * @param tf (unused)
   * @param global_costmap (unused)
   * @param local_costmap A pointer to the local_costmap used by the navigation stack
   */
  void initialize(std::string name, tf2_ros::Buffer*,
                  costmap_2d::Costmap2DROS*, costmap_2d::Costmap2DROS* local_costmap);

  /**
   * @brief  Run the DetectRecovery recovery behavior.
   */
  void runBehavior();

  /**
   * @brief  Destructor for the detect recovery behavior
   */
  ~DetectRecovery();

private:
  costmap_2d::Costmap2DROS* local_costmap_;
  bool initialized_;
  double sim_granularity_, min_rotational_vel_, max_rotational_vel_, acc_lim_th_, tolerance_, frequency_;
  base_local_planner::CostmapModel* world_model_;
  bool canGofoword,canGoback;
  double pi;
  bool turnAngle(double angle);
  void scanCallback(const sensor_msgs::LaserScan::ConstPtr& msg);
  void publishZeroVelocity();
  void goStraight(double distance);
  void goBack(double distance);
};
};  // namespace detect_recovery
#endif  // DETECT_RECOVERY_DETECT_RECOVERY_H


src/detect_recovery.cpp
/*********************************************************************
*
* Software License Agreement (BSD License)
*
*  Copyright (c) 2020, Anchuanxu.
*  All rights reserved.
* Author: Chuanxu An
*********************************************************************/
#include <detect_recovery/detect_recovery.h>
#include <pluginlib/class_list_macros.h>
#include <nav_core/parameter_magic.h>
#include <tf2/utils.h>
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Point.h>
#include <angles/angles.h>
#include <algorithm>
#include <string>
#include <std_msgs/Int16MultiArray.h>
#include <tf/transform_listener.h>
#include <cmath>


// register this planner as a RecoveryBehavior plugin
PLUGINLIB_EXPORT_CLASS(detect_recovery::DetectRecovery, nav_core::RecoveryBehavior)

namespace detect_recovery
{
DetectRecovery::DetectRecovery()
{
  ;
}

//void DetectRecovery::initialize(std::string scan_topic, std::string odom_topic)
void DetectRecovery::initialize(std::string name, tf2_ros::Buffer*,
                                costmap_2d::Costmap2DROS*, costmap_2d::Costmap2DROS* local_costmap)
{
  nh_ = new ros::NodeHandle("~/");
  laser_sub_ = nh_->subscribe<sensor_msgs::LaserScan>("/scan",1,boost::bind(&DetectRecovery::scanCallback,this,_1));
  odom_sub_ = nh_->subscribe<nav_msgs::Odometry>("/odom",1,boost::bind(&DetectRecovery::odomCallback,this,_1));
  vel_pub_ = nh_->advertise<geometry_msgs::Twist>("/cmd_vel",1);
  base_frame_ = "base_link";
  laser_frame_ = "laser";
  nh_->setParam("detect_recovery/base_frame", base_frame_);
  nh_->setParam("detect_recovery/laser_frame", laser_frame_);
  laser_sub_.shutdown();
  odom_sub_.shutdown();

  getLaserTobaselinkTF(laser_frame_, base_frame_);
  move_base_cancel_ = false;
  pi = 3.14;
  frequency_ = 10;
  initialized_ = true;
}

DetectRecovery::~DetectRecovery()
{
  if(world_model_)
  {
    delete world_model_;
    world_model_ = nullptr;
  }
  if(nh_)
  {
    delete nh_;
    nh_ = nullptr;
  }
}

void DetectRecovery::runBehavior()
{
    ROS_INFO("---Detect Recovery started---");


    laser_sub_ = nh_->subscribe<sensor_msgs::LaserScan>("/scan",1,boost::bind(&DetectRecovery::scanCallback,this,_1));
    odom_sub_ = nh_->subscribe<nav_msgs::Odometry>("/odom",1,boost::bind(&DetectRecovery::odomCallback,this,_1));

    ros::NodeHandle movebaseCancel_nh;
    ros::Subscriber movebasecancel_sub;
    movebasecancel_sub = movebaseCancel_nh.subscribe<actionlib_msgs::GoalID>("/move_base/cancel",1,
                                        boost::bind(&DetectRecovery::movebaseCancelCallback,this,_1));
    //如果有中断请求，则停止恢复
    if(move_base_cancel_ == true)
    {
      publishZeroVelocity();
      ROS_INFO("---move_base canceled---");
      move_base_cancel_ = false;

      laser_sub_.shutdown();
      odom_sub_.shutdown();
      return;
    }

    //从里程计获取现在的位置和姿态
    nav_msgs::Odometry odom_data;
    ros::Duration(0.5).sleep();
    this->getOdomData(odom_data);
    double robot_start_x = odom_data.pose.pose.position.x;
    double robot_start_y = odom_data.pose.pose.position.y;
    double robot_start_t = tf::getYaw(odom_data.pose.pose.orientation);
    uint32_t count_step = 0;
    while(isnan(robot_start_t))
    {
       ROS_INFO("can't get right robot post");
       if(count_step++ > 10)
       {
         laser_sub_.shutdown();
         odom_sub_.shutdown();
         return;
       }
       ros::Duration(1).sleep();
       this->getOdomData(odom_data);
       robot_start_x = odom_data.pose.pose.position.x;
       robot_start_y = odom_data.pose.pose.position.y;
       robot_start_t = tf::getYaw(odom_data.pose.pose.orientation);
    }
    //定义现在的位置和姿态
    double robot_current_x = robot_start_x;
    double robot_current_y = robot_start_y;
    double robot_current_t = robot_start_t;


    //获取激光数据
    std::vector< std::pair<double,double> > laser_point;
    this->getLaserPoint(laser_point);//使用回调函数赋值
    count_step=0;
    while(laser_point.size() == 0) //如果激光没有数据，等待10s，如果还没有，退出恢复
    {
      ROS_INFO("can't get scan data");
      if(count_step++ > 10)
      {
        laser_sub_.shutdown();
        odom_sub_.shutdown();
        return;
      }
      ros::Duration(1).sleep();
      this->getLaserPoint(laser_point);//使用回调函数赋值
    }
    size_t lase_data_size = laser_point.size();
    bool no_point = false;
    bool turn_half = false;
    bool turn_done = false;
    bool goStright = false;
    geometry_msgs::Twist cmd_vel;

    while(ros::ok() && !goStright){
    while(ros::ok() && turn_done == false)//如果还没有旋转一周，就可以开始恢复了。
    {
      this->getLaserPoint(laser_point);//获取最新激光数据
      size_t index = 0;
      for(index=0; index<lase_data_size; index++)
      {
        if(laser_point[index].first < 0.6 && fabs(laser_point[index].second) < 0.35) //截取前方60cm，左右35cm的数据
          break;//发现障碍物，直接跳出遍历循环。
      }
      if(index == lase_data_size)//如果顺利扫描一圈，则说明没有障碍物
      {
        no_point = true;
      }
      cmd_vel.angular.z = no_point == true?0:0.25;//没有障碍物，停下机器人，有就以0.25的速度继续旋转
      this->vel_pub_.publish(cmd_vel);
      if(no_point == true)
        break;
      this->getOdomData(odom_data);
      robot_current_t = tf::getYaw(odom_data.pose.pose.orientation);
      if(fabs(robot_current_t - robot_start_t) > 2)
        turn_half = true;//旋转了一半
      if(turn_half == true && fabs(robot_current_t-robot_start_t) < 0.1)
      {
        //确实已经旋转了一半 而且角度差又很小，说明回到了原点。
        turn_done = true;
        cmd_vel.angular.z = 0;
        this->vel_pub_.publish(cmd_vel);//旋转一圈，停止旋转。
      }
      ros::Duration(0.05).sleep();
      ros::spinOnce();
    }
    //前进逻辑
    if(no_point == true)
    {
      this->getOdomData(odom_data);
      robot_current_x = odom_data.pose.pose.position.x;
      robot_current_y = odom_data.pose.pose.position.y;

      robot_start_x = robot_current_x;
      robot_start_y = robot_current_y;
      double diff_x = robot_current_x - robot_start_x;
      double diff_y = robot_current_y - robot_start_y;

      double distance = sqrt(diff_x*diff_x+diff_y*diff_y);
      while(ros::ok() && distance < 0.2)
      {
        //std::cout << "distance: " << distance << std::endl; //debug
        this->getLaserPoint(laser_point);
        size_t index = 0;
        for(index=0;index<laser_point.size();index++)
        {
          if(laser_point[index].first < 0.4 && fabs(laser_point[index].second) < 0.35)
          {
            cmd_vel.linear.x = 0;
            cmd_vel.angular.z = 0;
            this->vel_pub_.publish(cmd_vel);//前方又有了障碍物，先停止
            no_point = false;
            break;//跳出遍历循环
//            turn_done = false;
//            turn_half = false;
          }
        }
        if(!no_point)
        {
          break;
        }
        //再跳出直行循环；

        cmd_vel.linear.x = 0.08;
        cmd_vel.angular.z = 0;
        this->vel_pub_.publish(cmd_vel);

        //更新位置
        this->getOdomData(odom_data);
        robot_current_x = odom_data.pose.pose.position.x;
        robot_current_y = odom_data.pose.pose.position.y;
        diff_x = robot_current_x - robot_start_x;
        diff_y = robot_current_y - robot_start_y;

        distance = sqrt(diff_x*diff_x+diff_y*diff_y);
        //std::cout << "+++distance: " << distance << std::endl; //debug
        if(distance >= 0.2)
        {
          goStright = true;
        }
        ros::Duration(0.05).sleep();
        ros::spinOnce();
      }
      cmd_vel.linear.x = 0;
      cmd_vel.angular.z = 0;
      this->vel_pub_.publish(cmd_vel);
    }
    }
    ROS_INFO("detect recovery end");
    laser_sub_.shutdown();
    odom_sub_.shutdown();
    return;
}
void DetectRecovery::odomCallback(const nav_msgs::Odometry::ConstPtr &msg)
{
  ROS_INFO_ONCE("odom data recevied");
  boost::mutex::scoped_lock lock(this->odom_mutex_);
  this->odom_data_ = *msg;
}
void DetectRecovery::movebaseCancelCallback(const actionlib_msgs::GoalID::ConstPtr &msg)
{
  move_base_cancel_ = true;
  return;
}
void DetectRecovery::getOdomData(nav_msgs::Odometry &data)
{
  data = this->odom_data_;
}
void DetectRecovery::getLaserData(sensor_msgs::LaserScan &data)
{
  data = this->laser_data_;
}
void DetectRecovery::getLaserPoint(std::vector<std::pair<double, double> > &data)
{
  data = this->point_vec_;
}
void DetectRecovery::publishZeroVelocity()
{
  geometry_msgs::Twist cmd_vel;
  cmd_vel.linear.x = 0;
  cmd_vel.linear.y = 0;
  cmd_vel.angular.z = 0;
  this->vel_pub_.publish(cmd_vel);
}
void DetectRecovery::getLaserTobaselinkTF(std::string laser_frame, std::string base_frame)
{
  tf::TransformListener* tf_ = new tf::TransformListener(ros::Duration(10));
  while(ros::ok())
  {
    if(tf_->waitForTransform(base_frame, ros::Time::now(), laser_frame, ros::Time::now(), laser_frame, ros::Duration(1)))
    {
      tf_->lookupTransform(base_frame,laser_frame,ros::Time::now(),transform);
      break;
    }
    ROS_WARN("frame %s to %s unavailable",base_frame.c_str(),laser_frame.c_str());
    ros::Duration(0.5).sleep();
  }
  //debug
//  std::cout << "laserTbase:" << std::endl;
//  std::cout << "Tx:" << transform.getOrigin().getX() << std::endl;
//  std::cout << "Ty:" << transform.getOrigin().getY() << std::endl;
//  std::cout << "Tz:" << transform.getOrigin().getZ() << std::endl;
//  std::cout << "Rx:" << transform.getRotation().getX() << std::endl;
//  std::cout << "Ry:" << transform.getRotation().getY() << std::endl;
//  std::cout << "Rz:" << transform.getRotation().getZ() << std::endl;
//  std::cout << "Rw:" << transform.getRotation().getW() << std::endl;

  delete tf_;
  tf_ = nullptr;
}
void DetectRecovery::scanCallback(const sensor_msgs::LaserScan::ConstPtr& msg)
{
  ROS_INFO_ONCE("scan data recevied");
  boost::mutex::scoped_lock lock(this->laser_mutex_);
  laser_data_ = *msg;
  this->point_vec_.clear();
  double min_angle = laser_data_.angle_min;
  double max_angle = laser_data_.angle_max;
  double laser_angle_increment = laser_data_.angle_increment; //雷达数据的角度间隔
  size_t laser_point_size = laser_data_.ranges.size();//雷达数据一帧的点数
  double theta = 0; //转角置为０；

  std::pair<double,double> tem_pair;
  for (size_t i = 0; i < laser_point_size; i++) {
    //做一些必要的判断，筛除无用的点
    if(isfinite(laser_data_.ranges[i]) == false)
      continue;
    if(isinf(laser_data_.ranges[i]) == true)
      continue;
    if(isnan(laser_data_.ranges[i]) == true)
      continue;
    tf::Quaternion q = transform.getRotation();
    theta = min_angle + i * laser_angle_increment + tf::getYaw(q);
    theta = normalizeAngle(theta,-M_PI,M_PI);
    if(theta >= -M_PI/2 && theta <= M_PI/2)
    {
     if(laser_data_.ranges[i] > 0)
     {
       tem_pair.first = laser_data_.ranges.at(i) * cos(theta) + transform.getOrigin().x();
       tem_pair.second = laser_data_.ranges.at(i) * sin(theta) + transform.getOrigin().y();
       this->point_vec_.push_back(tem_pair);
     }
    }
  }
}
};  // namespace detect_recovery
